pkg/ipam/allocate.go

```diff
@@ -98,6 +113,60 @@ func (i *ipam) Allocate(ctx context.Context, addArgs *models.IpamAddArgs) (*mode
 	return addResp, nil
 }
 
+func (i *ipam) releaseStsOutdatedIPIfNeed(ctx context.Context, addArgs *models.IpamAddArgs,
+	pod *corev1.Pod, endpoint *spiderpoolv2beta1.SpiderEndpoint, podTopController types.PodTopController) (bool, error) {
+	logger := logutils.FromContext(ctx)
+
+	preliminary, err := i.getPoolCandidates(ctx, addArgs, pod, podTopController)
+	if err != nil {
+		return false, err
+	}
+	poolMap := make(map[string]map[string]struct{})
+	for _, candidates := range preliminary {
+		if _, ok := poolMap[candidates.NIC]; !ok {
+			poolMap[candidates.NIC] = make(map[string]struct{})
+		}
+		pools := candidates.Pools()
+		for _, pool := range pools {
+			poolMap[candidates.NIC][pool] = struct{}{}
+		}
+	}
+	endpointMap := make(map[string]map[string]struct{})
+	for _, ip := range endpoint.Status.Current.IPs {
+		if _, ok := endpointMap[ip.NIC]; !ok {
+			endpointMap[ip.NIC] = make(map[string]struct{})
+		}
+		if ip.IPv4Pool != nil && *ip.IPv4Pool != "" {
+			endpointMap[ip.NIC][*ip.IPv4Pool] = struct{}{}
+		}
+		if ip.IPv6Pool != nil && *ip.IPv6Pool != "" {
+			endpointMap[ip.NIC][*ip.IPv6Pool] = struct{}{}
+		}
+	}
+	if !checkNicPoolExistence(endpointMap, poolMap) {
+		logger.Sugar().Infof("nic pool not exist, endpoint map: %v, pool map: %v", endpointMap, poolMap)
+		if endpoint.DeletionTimestamp == nil {
+			logger.Sugar().Infof("delete outdated endpoint of statefulset pod: %v/%v", endpoint.Namespace, endpoint.Name)
+			if err := i.endpointManager.DeleteEndpoint(ctx, endpoint); err != nil {
+				return false, err
+			}
+		}
+		err := i.release(ctx, endpoint.Status.Current.UID, endpoint.Status.Current.IPs)
+		if err != nil {
+			return false, err
+		}
+		logger.Sugar().Info("remove outdated of StatefulSet pod %s/%s: %v", endpoint.Namespace, endpoint.Name, endpoint.Status.Current.IPs)
+		if err := i.endpointManager.RemoveFinalizer(ctx, endpoint); err != nil {
+			return false, fmt.Errorf("failed to clean statefulset pod's Endpoint when expected ippool was changed: %v", err)
+		}
+		endpoint = nil
+		return true, nil
+	} else {
+		logger.Sugar().Debugf("statefulset Pod: owned pool %v, expected pools: %v", endpointMap, poolMap)
```

Comment:
logger.Sugar().Debugf("statefulset Pod does not need to release IP: owned pool %v, expected pools: %v", endpointMap, poolMap)